api/envoy/api/v2/core/health_check.proto

```diff
@@ -163,6 +163,18 @@ message HealthCheck {
     }
   }
 
+  // Health checks occur over the transport socket specified for the cluster. This implies that if a
+  // cluster is using a TLS-enabled transport socket, the health check will also occur over TLS.
+  //
+  // This allows overriding the cluster TLS settings, just for health check connections.
+  message TLSOptions {
+    // Specifies the ALPN protocols for health check connections. This is useful if the
+    // corresponding upstream is using ALPN-based :ref:`FilterChainMatch
+    // <envoy_api_msg_listener.FilterChainMatch>` along with different protocols for health checks
+    // versus data connections. If empty, no ALPN protocols will be set on health check connections.
+    repeated string alpn_protocols = 1;
```

Comment:
Not sure how much to bikeshed here, but we could imagine making this a full TLS context and specifying it has merge override semantics on the actual TLS socket.

Reply:
Hmm I guess that could be useful along with health check configs that use `alt_port`...  However we'd need a much bigger refactoring, _I think_, to be able to plumb through a full `UpstreamTlsContext` and be able to override the `TransportSocketFactory` settings...

Reply:
Yeah we should probably punt on this, unless someone feels strongly about it.

Reply:
I agree, I'm willing to go with what you have here if @mattklein123 and @PiotrSikora agree, but I think it would be nice if we had a clearer picture long term on how we want to handle TLS overrides. Clusters already have a dedicated scheme for this with transport socket selectors.

Reply:
This seems fine to me for now from an API perspective.

Reply:
Cool thanks -- resolving.